modules/thermal_hydraulics/src/userobjects/FlowChannelHeatStructureCouplerUserObject.C

```diff
@@ -0,0 +1,177 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "FlowChannelHeatStructureCouplerUserObject.h"
+#include "MooseMesh.h"
+#include "KDTree.h"
+#include "Assembly.h"
+#include "metaphysicl/parallel_dualnumber.h"
+#include "metaphysicl/parallel_numberarray.h"
+#include "metaphysicl/parallel_semidynamicsparsenumberarray.h"
+
+InputParameters
+FlowChannelHeatStructureCouplerUserObject::validParams()
+{
+  InputParameters params = ElementUserObject::validParams();
+  params.addRequiredParam<FlowChannelAlignment *>("_fch_alignment",
+                                                  "Flow channel alignment object");
+  params.addClassDescription(
+      "Base class for caching quantities computed between flow channels and heat structures.");
+  return params;
+}
+
+FlowChannelHeatStructureCouplerUserObject::FlowChannelHeatStructureCouplerUserObject(
+    const InputParameters & parameters)
+  : ElementUserObject(parameters),
+    _fc_elem_id(0),
+    _hs_elem_id(0),
+    _fc_qp(0),
+    _hs_qp(0),
+    _fch_alignment(*getParam<FlowChannelAlignment *>("_fch_alignment")),
+    _elem_qp_map(buildQuadraturePointMap())
+{
+}
+
+void
+FlowChannelHeatStructureCouplerUserObject::initialize()
+{
+  const auto map_ptrs = getCachedQuantityMaps();
+  for (const auto map_ptr : map_ptrs)
+    map_ptr->clear();
+}
+
+void
+FlowChannelHeatStructureCouplerUserObject::execute()
+{
+  unsigned int n_qpts = _qrule->n_points();
+  if (_current_elem->processor_id() == this->processor_id())
+  {
+    _fc_elem_id = _current_elem->id();
+    _hs_elem_id = _fch_alignment.getNearestElemID(_current_elem->id());
+
+    const auto map_ptrs = getCachedQuantityMaps();
+    for (const auto map_ptr : map_ptrs)
+    {
+      (*map_ptr)[_fc_elem_id].resize(n_qpts);
+      (*map_ptr)[_hs_elem_id].resize(n_qpts);
+    }
+
+    for (_fc_qp = 0; _fc_qp < n_qpts; _fc_qp++)
+    {
+      _hs_qp = _elem_qp_map.at(_fc_elem_id)[_fc_qp];
+      computeQpCachedQuantities();
+    }
+  }
+}
+
+void
+FlowChannelHeatStructureCouplerUserObject::threadJoin(const UserObject & uo)
+{
+  const auto & fc_hs_uo = static_cast<const FlowChannelHeatStructureCouplerUserObject &>(uo);
+
+  const auto map_ptrs = getCachedQuantityMaps();
+  const auto other_map_ptrs = fc_hs_uo.getCachedQuantityMaps();
+
+  for (unsigned int i = 0; i < map_ptrs.size(); ++i)
+    for (auto & it : *other_map_ptrs[i])
+      (*map_ptrs[i])[it.first] = it.second;
+}
+
+void
+FlowChannelHeatStructureCouplerUserObject::finalize()
+{
+  const auto map_ptrs = getCachedQuantityMaps();
+  for (const auto map_ptr : map_ptrs)
+    allGatherMap(*map_ptr);
+}
+
+const std::vector<ADReal> &
+FlowChannelHeatStructureCouplerUserObject::getCachedQuantity(
+    dof_id_type elem_id,
+    const std::map<dof_id_type, std::vector<ADReal>> & elem_id_to_values,
+    const std::string & description) const
```

Comment:
Why do you need the description here?

Reply:
It's just used in that error message. I can just use some generic error message instead though.

Reply:
This is fine, it provides more information than a generic message. I was just afraid I had missed something.